Fix all things snackbars

oakkitten opened the pull request:
This makes snackbars look like snackbars, behave like snackbars, move the way you would expect them to move. No surprises, nothing fancy, just plain ol' boring snackbars.

<img src="https://user-images.githubusercontent.com/1710718/178287362-7431d8cf-b008-4ebd-b3db-80ec67246260.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287372-c8793c3f-bc81-4ebd-ac10-e55c22cc4ad0.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287383-be02a320-e612-4124-baad-2f86597cffa6.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287402-c0f271ae-3113-4069-b223-04416edd8391.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287419-74e8920b-6822-4295-9456-569093218943.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287447-8e054c02-9cf9-464b-b463-469ce35d47e8.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287460-9f72fd98-a42b-4d72-8a94-7cb705b57d08.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287475-15e63d44-54e3-438c-884f-911a5d9e7122.png" width=250><img src="https://user-images.githubusercontent.com/1710718/178287486-c2b558be-17a2-4ed6-afc6-fae8f9d849b6.png" width=250>

## Purpose / Description

Currently, there are many problems with snackbars.

  * Snackbars can appear too low (inside the gesture zone), too high (some distance above the gesture zone), and *way* too high, over toolbar (huh??);
  * Snackbars don't visibly move when you drag them;
  * Snackbars can only be flicked or dragged to the right;
  * When you dismiss the snackbar by flicking, the fab will get stuck;
  * Snackbars aren't properly styled. They are rectangularey, no margins, no elevation;
  * When insets change, snackbars fail to adjust position properly.

You can see these problems in many an app, including the ones from Google itself. Most of these issues arise from the bugs in the material library. These bugs mask other bugs in material library:

  * When you dismiss a snackbar, it will be moved to the side, but that ignores snackbar margins, so it will briefly appear stuck near the edge;
  * Flicking a snackbar may dismiss it in the wrong direction;
  * `CoordinatorLayout` updates may cause snackbar flicker in original position while it is settling (found [this](https://stackoverflow.com/questions/22865291) after fixing)
  * This...

https://user-images.githubusercontent.com/1710718/178286944-ab51d05b-777f-43ee-92e6-cf5fdc839a5b.mp4

## Approach

I tried to explain neatly in every commit what it addresses and how. Please take a look!

I resorted to some pretty awful workarounds in some places. While what they do may be atrocious, I tried making these workarounds in such a way that it's easy to remove them if need be.

## How Has This Been Tested?

Tested on my devices running Android 11 and 6. I can not run emulator, so this was **NOT** tested on other devices. I hope that this **IS** tested on more devices before this is merged.

## Some thoughts

* Snackbars are “supposed” to appear above fab according to the guidelines, but our fab expands so that wouldn't work I guess. Besides, in Gmail, which is I guess a paragon of material design or something, snackbars also move the fab. ¯\\\_(ツ)_/¯
* We could make snackbar color reflect its purpose, e.g. green snackbars for success, red for errors.
* Some app allow long-tapping on snackbars to get some debug level information, such as full exception tracebacks. I think this is pretty intuitive!

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

mikehardy commented:
Wow! This is awesome. Apparently it already took conflicts because...also awesome. But your description had me chuckling #thanksgoogle right?

Arthur-Milchior commented:
How did you end up with your video? I assume this is not the end result you want (because clearly, dismissing the fab or the deck picker is not anyone's goal). But it's also not something that can be done today. I mean, I didn't check, but I'm certain it would already have been notified to us if it were the case

Arthur-Milchior commented:
If you create multiple snackbar right now, the FAB moves up multiple time without ever going down. Not a huge deal, but if you have a way to deal with it, it would seems nicer.
It's not currently an issue with main because the current snackbar disappear before the next appear

Arthur-Milchior commented on AnkiDroid/src/main/res/values/attrs.xml:
```diff
@@ -129,6 +129,9 @@
     <attr name="editTextDisabled" format="color"/>
 
+    <attr name="snackbarBackgroundTintColor" format="color"/>
+    <attr name="snackbarButtonTextColor" format="color"/>
```
Can you please add in the commit message that you ran the accessibility tool and ensured just that colors passed it succesfully.

oakkitten replied:
Actually I only know realized that I didn't change all themes. In fact the accessibility scanner does complain about the button text color fo both dark and black themes. Fixed in 2ea2f760, also finally found why button text colors were not exactly the ones specified in the theme and fixed that too.

Arthur-Milchior commented on AnkiDroid/src/main/res/values/styles.xml:
```diff
@@ -166,4 +166,16 @@
+        <item name="android:textColor">?attr/snackbarButtonTextColor</item>
+        <!-- Material buttons have some letter spacing added, but since no other components
+             in the app use material theming, let's remove it for the sake of consistency.
+             TODO Remove once app is using material theming. -->
```
I'd have stated "if the app move to material theming", because it is not obvious to me it will occur

oakkitten replied:
Fixed in 23e83399

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -30,54 +30,59 @@ import com.ichi2.anki.UIUtils
-                mTranslationY = translationY
-            }
+        if (dependency is SnackbarLayout && dependency.visibility == View.VISIBLE) {
+            val y = getFabTranslationYForSnackbar(fab, dependency)
```
I did love the value name `translationY`, just to note that it's a translation value and not an absolute value

oakkitten replied:
Fixed in 7a6734ae, also changed a few other names

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -30,54 +30,59 @@ import com.ichi2.anki.UIUtils
+private const val SNACKBAR_FADE_IN_DURATION = 150L
+private const val SNACKBAR_FADE_OUT_DURATION = 75L
+
+/**
```
I don't see the point of changing // to /*

Not that it's bad. Just that it is shown as deletion on the reviewer tool, and not entirely fan of needless change in an already huge PR

oakkitten replied:
Sorry! (I blame `git absorb`, sometimes it picks up the wrong commit to fixup, I think this was it.)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -30,54 +30,59 @@ import com.ichi2.anki.UIUtils
+ *   * [BaseTransientBottomBar.ANIMATION_FADE_IN_DURATION]
+ *   * [BaseTransientBottomBar.ANIMATION_FADE_OUT_DURATION]
+ */
+private const val SNACKBAR_FADE_IN_DURATION = 150L
```
Implementor/other reviewer choice.
```
private val SNACKBAR_FADE_IN_DURATION: Long = BaseTransientBottomBar::class.java.getDeclaredField("ANIMATION_FADE_IN_DURATION").let{
    it.isAccessible = true
    val duration = it.getLong(null)
    it.isAccessible = false
    return@let duration
}
```
On the plus side, it's hard coded as the same constant. 
On the minus side, each time a FabBehavior is used, it must do this reflection again

oakkitten replied:
It's not exactly important that the duration is the same. It's so fast and for both the snackbar and the fab it's start-fast-end-slow so I don't think it would be noticeable much if it's off by a bit. Besides, the fade-out animations are not run simultaneously. Don't think reflection is worth it. (Ugh, reflection.)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -30,54 +30,59 @@ import com.ichi2.anki.UIUtils
-                fab.translationY = translationY
-                mTranslationY = translationY
-            }
+        if (dependency is SnackbarLayout && dependency.visibility == View.VISIBLE) {
```
I'd love to understand why it was tested whether fab is visible and now whether the dependency is visible. this change si not clear to me here

oakkitten replied:
You know why you are a great reviewer? I don't think I would ask this question if I was reviewing. Visibility check? Well it probably fixes... *something*, couldn't hurt? 

Well, I *think* it was at some point fixing something. Maybe something with insets. I didn't remember so I went to check what happens if I remove the check. A-and this. Fixed in 92f7d7cc

https://user-images.githubusercontent.com/1710718/179000297-89a16f85-161a-4c93-8426-58c96778cb54.mp4

(Hopefully what it was meant to fix is no longer an issue. At any rate, I tested around and couldn't find any issues.)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -30,54 +30,59 @@ import com.ichi2.anki.UIUtils
-                fab.translationY = 0f
-                mTranslationY = 0f
-            }
+        translateFabToY(fab, 0f, SNACKBAR_FADE_OUT_DURATION)
```
I don't understand why you need to have a check for `onDependentViewChanged` but not `removed`. 

Given `layoutDependsOn` I'd assume the "instance of" `is` test is useless there. But then it would make sens to remove in both place. Also, are we sure there is no concurrency issue. I could imagine that, if there is two successive snackbar, the method to note the first one changed is called first, and the method to notify the first one is removed is called second. So you would not be translating in the right order

Arthur-Milchior reviewed (commented):
I'm going to stop reviewing for today here, already quite long.

I'm in this strange situation where I don't know exactly how to reply, because, on the one hand, I've seen you around for quite some time. On the other, it's the first time I review your code. So you risks to have already heard multiple time what I will state below.

Feel free to disagree with any comment I made, you seems more knowledgeable than myself in UI and I may have misunderstood things here and there. I'd just would love that if you decide not to apply a suggestion I made, to answer to it to explain why. 


Thanks a lot. Screenshot is nice. I never felt that the appearance or not being able to slide out is such an issue, but I guess it's a nice improvement, especially if you want to click on some deck that is hidden by the snackbar and you have to wait for it to disappear otherwise.

Your comment makes it clear that there is nothing to see here but ordinary snackbar. And I love that. Applying the principles of less surprises. Following well established standards everywhere that is not a justified specificity of AnkiDroid.
However, it fails to explain to me why it was not the case with current snack bars. They also seemed to me like adequate snackbar. If you could also add a screenshot of what was the issue with current implementation of snack bar, it would help understand. I must admit that I personally never noticed a real issue. Admittedly, I never tried to discard it manually before. 

To ask my question in another way. Are the only changes related to the appearance, unstucking the FAB, and having a discarding animation? Or am I missing something (maybe, I've still plenty of commits to read)



I also want to thank you for the atomiticy of your commit. Makes for an easy review of an otherwise hard PR (I'm not used to UI, and don't know relevant libraries). I do appreciate the style of the first commit's message, I find it a nice touch

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -61,21 +63,21 @@ class FabBehavior : CoordinatorLayout.Behavior<View> {
+private val MAX_OVERLAP = UIUtils.convertDpToPixel(10f, AnkiDroidApp.getInstance().applicationContext)
+
+private fun getFabTranslationYForSnackbar(parent: CoordinatorLayout, fab: View): Float {
+    return parent.getDependencies(fab)
```
I could, but I prefer to only use the shorthand way only with simple expressions. This way I don't have to break line around the `=` or after `parent`, both of which are a bit ugly, and it's immediately clear what the return type is, and if I ever want to change the method, which I in fact do, this leads to a more sensible diff. I think

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -56,7 +56,7 @@ class FabBehavior : CoordinatorLayout.Behavior<View> {
         if (dependency is SnackbarLayout && fab.visibility == View.VISIBLE) {
             val translationY = getFabTranslationYForSnackbar(parent, fab)
-            if (translationY == mTranslationY) {
+            if (translationY != mTranslationY) {
```
My commits are not signed, so feel free to cherry-pick!

(However I like to keep related commits together)

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SwipeDismissBehaviorFix.kt:
```diff
@@ -0,0 +1,81 @@
+ * and ask parent to stop intercepting further touches. This propagates to CoordinatorLayout,
+ * which then resets touch behavior for its children--including us.
+ *
+ * The sequence of unfortunate events:
```
These events still occur. This is just an exact call stack that's described in the previous paragraph. This fix does not prevent the events from happening, but does prevent them from doing harm.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SwipeDismissBehaviorFix.kt:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    override fun onTouchEvent(parent: CoordinatorLayout, child: V, event: MotionEvent): Boolean {
```
I would absolutely NOT like `= !ignoreCallsToOnTouchEvent && super.onTouchEvent(parent, child, event)` 😛. Yes, the logic is the same, but it doesn't read well. My varint, with the guard clause, reads

> If flag to ignore this call is set, abort.
> Proceed to call super.

While the shortcut variant reads

> Return whether the flag to ignore this call is not set and at the same time super returns true.

Does this make sense to you? 🙈

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SwipeDismissBehaviorFix.kt:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    override fun onTouchEvent(parent: CoordinatorLayout, child: V, event: MotionEvent): Boolean {
```
I *feel* like this should be evident from the code? At the beginning of `onInterceptTouchEvent` a flag `ignoreCallsToOnTouchEvent` is set, and on end it is reset, and checked in `onTouchEvent`. That's all the code does! 

...or rather it did at the time when I wrote that commit. Since it has one additional line now, I guess I can add a clarification. Added in fab88082

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/UIUtils.kt:
```diff
@@ -122,15 +120,10 @@ object UIUtils {
         }
-        // Make the text white to avoid interference from our theme colors.
-        val view = sb.view
-        val tv = view.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
```
B-but that code was deleted. It wasn't even mine...

oakkitten commented on AnkiDroid/src/main/res/values/theme_dark.xml:
```diff
@@ -123,6 +123,11 @@
 
         <item name="android:navigationBarColor">@color/black</item>
+
+        <!-- Snackbar. snackbarStyle, snackbarTextViewStyle and snackbarButtonStyle must be present -->
```
You are totally right, awful comment. Clarified in 5b402c71

oakkitten commented:
> How did you end up with your video? I assume this is not the end result you want

This is the problem that you begin to see when you fix the problem with snackbars not moving at all. They don't move since they fail to properly intercepted touch events, and hence the view is never captured.

adcc5e0 fixes the issue in the video; you should be able to see the bug if you revert it.

> If you create multiple snackbar right now, the FAB moves up multiple time without ever going down. Not a huge deal, but if you have a way to deal with it, it would seems nicer.

Fixed in 92f7d7cc thanks to another comment of yours! 
![image](https://user-images.githubusercontent.com/1710718/179002415-4183ebad-f2e0-4082-a5fa-890f19bc56aa.png)


> I'm in this strange situation where I don't know exactly how to reply

Feel free to treat me as if I learned Java yesterday. I make silly mistakes and with rebasing—and there was a lot of rebasing here—they tend to creep in, especially in the form of less than optional approaches and ill-fitting names. And when I do a self-review, sometimes it's hard to escape thinking in terms of the old code. Oof!

> However, it fails to explain to me why it was not the case with current snack bars. They also seemed to me like adequate snackbar.

Well, it is, if you ignore the fact that they sometimes appear inside the gesture zone, or over the fab, or over the toolbar, or make the fab stuck, or don't move when you swipe them 😛. None of these issues are exactly super important. Just something I decided to fix.

> To ask my question in another way. Are the only changes related to the appearance, unstucking the FAB, and having a discarding animation?

That's pretty much it, yeah. Nothing fancy. A-and a change in the way activities create snackbars.

Arthur-Milchior commented on AnkiDroid/src/main/res/values/dimens.xml:
```diff
@@ -28,4 +28,12 @@
     <dimen name="note_editor_toolbar_height">52dp</dimen>
+
+    <!--
+      The IDE gives the following warning for this dimen; I think this is a false positive?
```
Even if I rremove the `tools:ignore` I don't see this warning in the IDE nor when compiling. So I'm surprised here. 
Where did you see it? Could you screenshot. I'd love to understand

oakkitten replied:
Since then I upgraded to AS Dolphin (unrelated reason) and it doesn't produce this warning ¯\\\_(ツ)_/¯

david-allison replied:
I'd be happy to remove this comment + ignore

oakkitten replied:
Done

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -83,28 +83,37 @@ open class SensibleSwipeDismissBehavior<V : View> : SwipeDismissBehavior<V>() {
             }
 
+            fun onViewSettled() {
+                if (dismiss is Dismiss.DoDismiss) {
```
You could also uses an helper method in Dismiss, that returns whether it's Left/Right Dismiss.

oakkitten replied:
Now it's an enum, anyway a67498e261391559f254b9de15a8ce09c1b7cf60

Arthur-Milchior reviewed (commented):
I stop at fixup for todays. 

I'm really astonished by the quantity of code this requires.
Given the amount of work you did, I assume you checked to see whether there is not already an available library that we could just incorporate as a dependency, or at worse copy as-is. However, just in case of doubt, I prefer to ask for confirmation.

I'm happy that:
* it's easy to test and see it works
* the worst case is either a crash or a snackbar that don't work as expected, which are two acceptable behaviors (I mean, not great, but no data corruption either), because even if I read everything, some parts requires a lot of insights about the UI. And since you work with hacking internal methods and inheriting to change some behavior, I can easily imagine I'm missing some cases that would lead to bug

dae commented on AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:
```diff
@@ -20,28 +20,32 @@ import android.app.Activity
+ * TODO: This seems to be a copy of from [com.ichi2.async.catchingLifecycleScope],
+ *   perhaps put the two methods together?
+ *
+ * TODO: The try/except block here catches CancellationException, is this right?
```
Divyansh and I were working on independent PRs in parallel which each ended up adding this routine. My version at least is still in flux (https://github.com/ankidroid/Anki-Android/blob/5ff92403d224d7599997cd2712d6b641b4c75b97/AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt#L39 in latest PR), but once it settles down, settling on one would seem to make sense. The only reason it doesn't ignore cancellation exceptions is because I haven't tested any code where that would be an issue yet, and it didn't occur to me.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:
```diff
@@ -143,6 +143,7 @@ abstract class AbstractFlashcardViewer :
     private var mScrollingButtons = false
     private var mGesturesEnabled = false
     private var mLargeAnswerButtons = false
+    private var mAnswerButtonsPosition: String? = "bottom"
```
`mAnswerButtonsPosition` only stores the contents of `answerButtonsPosition` for later use. This *should* be an enum, but I suppose enum-ifying this is should be a part of the/a preferences PR?

BrayanDSO replied:
There is nothing to change on the preference itself, i.e. the interactive piece of UI that makes possible to change this setting, or on any part of its configuration. And this is only used on the AbstractFlashcardViewer. So, IMHO, this seems more like part of a "refactor AbstractFlashcardViewer" PR.

About the enum itself, seems like a good idea.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:
```diff
@@ -1765,20 +1765,13 @@ abstract class AbstractFlashcardViewer :
+
+        if (mAnswerButtonsPosition == "bottom") {
+            val easeButtons = findViewById<View>(R.id.answer_options_layout)
+            val previewButtons = findViewById<View>(R.id.preview_buttons_layout)
```
Don't microoptimize! Don't don't don't. Anyway, I thought AnkiDroid was moving to view binding, which should take care of this, wasn't it?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:
```diff
@@ -1855,27 +1853,21 @@ open class CardBrowser : NavigationDrawerActivity(), SubtitleListener, DeckSelec
-            val displayText: String = if (cardCount == 0) {
-                getString(R.string.card_browser_no_cards_in_deck, selectedDeckNameForUi)
+            if (hasSelectedAllDecks()) {
+                showSnackbar(subtitleText, Snackbar.LENGTH_SHORT)
```
I strongly prefer using guard clauses for, well, guarding. There's a misused guard clause just above this statement, and tbh it also shouldn't be there. This way the method *should* read:

    Update list.

    If have searched:
        If all decks are selected,
            show short snackbar;
        else,
            prepare message,
            show indefinite snackbar with message.

If using a return statement, this becomes,

    Update list.
    If have not searched, don't proceed. [0]

    If selected all decks,
        show short snackbar;
        don't proceed. [1]

    Prepare message, [2]
    show indefinite snackbar with message.

I don't like the latter part because:
    * This doesn't read well as text. It is not clear that [2] is an alternate route.
    * When you read [2] and [1], you have to be aware of [1] and [0]. In the former case, the flow is clear from the indentation level.
    * Having multiple return statements is generally an antipattern. Guard clauses are an exception because they are meant to exit the method *before* the primary body of it is run. You would generally place them even before the log statements.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:
```diff
@@ -20,28 +20,32 @@ import android.app.Activity
         } catch (exc: Exception) {
-            // TODO: localize
-            showSimpleSnackbar(activity, "An error occurred: {exc}", false)
+            val exceptionMessage = exc.localizedMessage ?: exc.toString()
```
The biggest problem with `"An error occurred: {exc}"` is that this literally displays `An error occurred: {exc}`... This is not Python! Other then that, the new `showSnackbar()` doesn't accept null text, so saying simply `exc.localizedMessage` doesn't work; also, I decided to add a case for when the error message is blank. This method is meant to catch *anything*, so it could result in snackbars with no text. This tries to show at least something useful in those cases.

Btw what's “factorize”?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/UIUtils.kt:
```diff
@@ -34,99 +29,6 @@ object UIUtils {
     }
 
-    /**
-     * Show a simple Toast-like Snackbar with no actions.
```
This *would* be a lot of work 🤯. But if the sets of arguments are the issue, I think you can rely on the compiler there as it won't allow putting wrong things in wrong places. 

Also, with the new API, apart from the builder, the only two arguments are:
* message (string or resource)
* length

So this should be easy, and the old API were consistent in the order of common arguments:
* showSimpleSnackbar:
   * activity
   * message (string or resource)
* showSnackbar:
   * activity
   * message
   * is short (boolean) OR length (int)
   * button text
   * button listener
   * view?
   * callback

As you see both methods start with activity and message. So these are basically the same method, with one having some defaults!

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/export/ActivityExportingDelegate.kt:
```diff
@@ -168,9 +168,11 @@ class ActivityExportingDelegate(private val activity: AnkiActivity, private val
-        val isSuccessful = exportToProvider(result.data!!, true)
-        @StringRes val message = if (isSuccessful) R.string.export_save_apkg_successful else R.string.export_save_apkg_unsuccessful
-        showSimpleSnackbar(activity, activity.getString(message), isSuccessful)
+        if (exportToProvider(result.data!!, true)) {
```
Fair point, fixed in b10f65086b4fadaaf181132caa0c1288f56c5fd4

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -0,0 +1,105 @@
+/*
+ * This program is free software; you can redistribute it and/or modify it under
```
Never understood the point of the copyright statement when you have many people who edit the same file. Doesn't everyone get a copyright? If so, then this statement is wrong for the vast majority of files, since not many people update it. Besides, don't I get the copyright regardless of the statement?

Also, what's the point of having a license header in every file? Having a project-wide license should cover every file that doesn't have a different explicit license and is not in the NOTES(?) file. (There are some files without the license header. What's up with that?)

mikehardy replied:
Could you peel this via context menu action to a new issue 🙏? I'm leaning your way these days on copyright. I'd rat have a license bot set up for PRs to sign a contributor license agreement (like invertase org, used in react-native-firebase), remove per file license lint check, and have a project wide license, with only the couple exceptions as needed. But it's a real, separate change assuming others agree, so needs debate / separate PR

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -0,0 +1,105 @@
+ *
+ *   * if calling on an activity, it **MUST** have a `CoordinatorLayout` with id root_layout;
+ *
+ *   * if calling on a view, it **MUST** be either a `CoordinatorLayout`,
```
> I'd appreciate that this sentence stands alone. So answer what must be either a coordinatorLayout or ...

Sorry I don't understand what you mean

> Also, you can probably use [CoordinatorLayout] so that it creates a link in Kotlin

Added full Javadoc in 63ac44b18d969b20f820d2bb4898a34f02a9c05c that has a link. (Btw it's annoying that you can say `` `CoordinatorLayout` `` and it's rendered in monospace font, but `[CoordinatorLayout]` is rendered in proportional, and you somehow can't do both?..)

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -0,0 +1,105 @@
+typealias SnackbarBuilder = Snackbar.() -> Unit
+
+/**
+ * You can create snackbars by calling `showSnackbar` on either an activity or a view.
```
Added full Javadoc in https://github.com/ankidroid/Anki-Android/commit/63ac44b18d969b20f820d2bb4898a34f02a9c05c

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -0,0 +1,105 @@
+ *     or a (possibly indirect) child of `CoordinatorLayout`.
+ *
+ * Any additional configuration can be done in the snackbar builder block, like below.
+ * This is preferred to call chaining, as it ensures that the show() method is called.
```
Fair point, removed the line in 4e4467afd9fe9fae3c0b1014f8cb5a3ac543acec

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -0,0 +1,105 @@
+
+    if (view != null) {
+        view.showSnackbar(text, duration, snackbarBuilder)
+    } else {
```
This is another one of those cases where I would prefer `if` branching because that's not a guard clause.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -0,0 +1,105 @@
+ * You can create snackbars by calling `showSnackbar` on either an activity or a view.
+ * As `CoordinatorLayout` is responsible for proper placement and animation of snackbars,
+ *
+ *   * if calling on an activity, it **MUST** have a `CoordinatorLayout` with id root_layout;
```
Fixed in e019ec4f0da816d704fde91a9f2cb947d8c736a1

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -92,14 +94,41 @@ fun View.showSnackbar(
+    addCallback(object : Snackbar.Callback() {
+        override fun onShown(snackbar: Snackbar) {
+            actualBehavior = SwipeDismissBehaviorFix<View>().apply {
+                listener = actualBehavior?.listener
```
This simply adds a callback, that, when the snackbar is shown, replaces its behavior with the fixed one. And while it does it, it steals the listener from the original behavior. And the listener is what deals with the timers.

Anyway, I went and revisited setting the behavior “the snackbar way” in 884c2d54659831f73181f573b99546f3560babc3, which removes this part. (Before, this would be “the coordinator layout way”. I tried that before but preferred the “the coordinator layout way” then. Given how many lines I got to delete with “the snackbar way” I like this better now. Oof, shame on me. But figuring this out was a long walk. This might look simple—and it is—but what I started with was a debugger and a complete mess of code.)

(P.S. What's a block? I think a block usually means the last lambda argument in Kotlin, Ruby also has a similar thing.)

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -0,0 +1,137 @@
+/*
```
I see your point, but I wouldn't want to write—or use—a library consisting mostly of workarounds... Hopefully material library fixes this one day, and some or all of these workarounds can be removed.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -0,0 +1,137 @@
+
+    /** See [com.google.android.material.behavior.SwipeDismissBehavior.dragCallback] */
+    inner class ViewDragHelperCallback : ViewDragHelper.Callback() {
+        private var initialChildLeft = Int.MIN_VALUE
```
Having `Int?` means we have to perform either nullability checks or say `!!`, and the former is ugly and the latter is somthing I like to avoid at all costs!

> In any case, I'd appreciate a javadoc to indicate it's an index relative to its parent.

Not sure I understand what's index relative to its parent. This is just, well, the initial left edge of child in pixels. Not sure if it matters whether it's relative to the parent.

> Also note somewhere we assume we have a single child view on which all methods are called

Well, we are, but at the same time we don't, in a way. This callback has almost no state, apart from the `initialChildLeft`, and in theory can work with other views. (By the way, the original callback sets a similar variable, `originalCapturedViewLeft`, on each capture. Which lead to snackbars getting stuck after you release them and catch them again while they are settling.) ~~It is `SwipeDismissBehaviorFix` that makes sure this is not getting other views.~~ As of b94c7ac5cd09be16af20b2dc0df941b557bd6453, this is the job of the callback (again). I hope it's more clear now that this only deals with snackbars.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -0,0 +1,137 @@
+         *
+         *   * Else do not dismiss, and return to original position.
+         */
+        private fun shouldDismiss(child: View, xvel: Float): Dismiss {
```
I have to confess I'm not very good at testing. 1a0945cd96a7e0c3d60551bb841dc2f1ec307d46 is my attempt. Hope it's decent.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -93,9 +98,13 @@ open class SensibleSwipeDismissBehavior<V : View> : SwipeDismissBehavior<V>() {
                     override fun run() {
                         if (viewDragHelper?.continueSettling(true) == true) {
                             child.postOnAnimation(this)
+                        } else if (dismiss is Dismiss.DoDismiss) {
```
I don't have a strong preference one way or another. Fixed in a67498e261391559f254b9de15a8ce09c1b7cf60

> Also, an equality comparison seems less expensive.

Do not microoptimize!!

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -131,3 +138,35 @@ private var Snackbar.actualBehavior: SwipeDismissBehavior<View>?
+ * and the one method is not private that calls it is [Snackbar.onAttachedToWindow],
+ * so we hack into its package namespace using a helper method.
+ */
+@RequiresApi(Build.VERSION_CODES.Q)
```
It is not actually a requirement, it's just that what this does is useless before Q since gestures weren't a thing back then

oakkitten commented:
> Given the amount of work you did, I assume you checked to see whether there is not already an available library that we could just incorporate as a dependency, or at worse copy as-is.

I did, not right away, but I did. It seems that there are quite a few libraries that make things that look like snackbars, but many of them have not been updated for quite a while and don't look material, and those that do look material often don't look right, and the readmes mention nothing about movement. Many of them come with bells and whistles, and I only wanted to fix what we had to behave right.

Besides, I prefer not to use libraries for things like this. You can see how `MaterialDialog` has been been problematic in AnkiDroid. In the end I often end up rooting out that library for one reason or another, and reimplementing the workarounds. I have to admit there are a bit too many of them here 😅. Well, hopefully a little fewer now (and probably don't count `SensibleSwipeDismissBehavior` as a hack).

> And since you work with hacking internal methods and inheriting to change some behavior, I can easily imagine I'm missing some cases that would lead to bug

All valid points. But you have to that a lot with Android, don't you. So many things are broken. I got used to this. 😔

P.S. Speaking of `fixMarginsWhenInsetsChange`, it is possible to rewrite the logic of whatever it calls without calling any internal methods. That logic isn't very great in the first place... Want me to do that? Otoh, at least it's not reflection.

mikehardy commented:
Have you considered a pseudo PR as patches to AOSP on Google issue tracker? Just thinking of any way to de-terrible the base code, even if slowly over time, in a proactive way vs hoping it magically improves. Might work?

mikehardy commented:
I'm generally positive on getting this in but I do need to review it all.

I followed the discussion / reviews / fixes as it happened and it seemed either well-defended or fixed+re-pushed

Tagging it as high priority because this thing has taken a *pile* of conflicts and it is wide-ranging enough I'd like it go in before that happens again

Arthur-Milchior commented:
> > And since you work with hacking internal methods and inheriting to change some behavior, I can easily imagine I'm missing some cases that would lead to bug
> 
> All valid points. But you have to that a lot with Android, don't you. 

Actually, I don't. In my experience, it has often been possible to use external library for our UIs - at least the part that I've seen done since I joined in 2020 - without having to do that many change to fix them. 
Admittedly, introduction slides required some extra work, but probably not as much as you did here

That may be why we got stuck with a bad snackbar. 

Outside of UI, I can recognize that JSON has been a pain, because every method sends exception. That was easy to fix. Transitioning to a more efficient fix however has failed  (and should soon become useless anyway)
Scoped storage however is a HUGE pain. But it's not like standard library are broken, it's more that Android didn't even both provide anything useful.

> P.S. Speaking of `fixMarginsWhenInsetsChange`, it is possible to rewrite the logic of whatever it calls without calling any internal methods. That logic isn't very great in the first place... Want me to do that? Otoh, at least it's not reflection.

I've no idea what you're speaking about. I can't find this `fixMarginsWhenInsetsChange` when looking at the code you changed.

If you refuse to rebase (you explained us why. Quoting linux. I can't force you. Even if this is not what we do here), can you at least do the merge with main, because there is a lot of conflict already, and I'd prefer if you could do it on your computer and push, than if we have to do it on github and then check whether it compiles.

Also, @oakkitten , did you see you had a question from Mike?

oakkitten commented:
Sorry, I have been a bit busy off the grid. Also, my cat have had babies and I have been busy worrying about the little fluffy sausages (even though Masha does an excellent job of caring for them herslf).

> Have you considered a pseudo PR as patches

I could but the issues with the snackbars are old, and have been reported, and they remain unfixed, so I just don't see that they want to fix this. One of the big mysteries with Google issues is their reasoning to fix or to ignore them (and the meaning of “adexe s nau”). Sometimes they fix a niche bug with no votes in a day, and sometimes they don't fix a major issue with hundreds of me toos in years. In my experience, pestering them does not work. Also, I suppose that by now many apps rely on the broken behavior to some extent and so they can't ✨ just fix ✨ it all. Also yet, I'm a tad fed up with snackbars for the time being and don't want to touch them in any way (a shame that I can't just turn them off system-wide in Android).

Anyway, if Google does fix their library, this PR makes it specifically easy to throw away workarounds. The rest of the changes such as the new API, which is the biggest change, or styling, or the fab animation, should still be good I hope!

> Actually, I don't.

I suppose that for the most part stuff works okay, it's the corner cases that might require workarounds. Some examples that I have fond memories of...

* EditTextPreference isn't (wasn't?) obtaining styles (such as `android:singleLine`) from the xml, so if you want that, you might want to subclass it it.
* On some Android versions, changing a DayNight theme does not (did not?) properly set task description color, so if you care about that, you might want to subclass your activity to do it.
* RecycleView does not take `setStackFromEnd` into account when doing `scrollToPositionWithOffset`, which results in slightly incorrect position restoration, so if you want it to be super precise you might want to override it.
* When scrolling to a position that is just outside of viewport, RecyclerView used to start scrolling too fast and would have to reverse. Again, this is not a big deal, you'd see this as a momentary glitch, but still used to be fixable by subclassing. (This took a lot of convinding, but somehow I managed to get Google to fix this.)
* This doesn't require subclassing, but this is an excellent example of a horrible fix that's required for a very corner case. If you are implementing bubbles, *and* you want to update notifications, you need to have a way to tell if a conversation linked to it will bubble or not, else you risk re-issuing a notification that the user has dismissed. This isn't possible to reliably do using the provided API. You can't normally detect a situation when the user has enabled *all* bubbles, but then disabled *some* of them. In this case, the only option you have to get the channel setting is to say `"mAllowBubbles=0" in notificationChannel.toString()`...... (Google has fixed one half of this problem in Android S and I'm trying to get them to fix this part as well. But even their own bubble demo app have a ton of issues with bubbles, including the mentioned problem with reappearing notifications, and you might have seen similar issues in other apps...)
* In some rare circumstances, tapping on an input field does not pop up the keyboard on the first tap. Mozilla [fixed it](https://bugzilla.mozilla.org/show_bug.cgi?id=1211848) by subclassing `BaseInputConnection`.
* While I prefer not to do this, in some cases getting to the bottom of rare crashes can be extremely hard. So sometimes overriding a method to call super while suppressing *this particular exception* is a viable option.

Well, it was fun remembering these hehe. Anyway, you see stuff like this is not super important to the functionaly of the apps. And it's not really worth to have a full library for this. Well, maybe it is for the preferences, as there are more issues with them, but then you get stuck with that library and the bugs that are in that library. I guess that snackbars, which have so many issues, can be covered by a library in a very nice way; you don't have to deeply entangle that library with your code as there's only a few entry points, and you can create your own facade wrapper and swap the underlying library every Wednesday if you want to. Yeah, but as for the other stuff... do you really want a full on RecyclerView alternative to fix a momentary visual glitch? You either ignore it, or subclass it and make it scroll one pixel to the direction of the scroll, and then to the position supplied by the caller. Scrolling even one pixel will layout enough off-screen children that RecyclerView will be able to properly measure the needed distance.

> I can't find this `fixMarginsWhenInsetsChange`

That's in Snackbars.kt. It calls `Snackbar.onAttachedToWindow()` by hacking into Snackbar's package space. If using GitHub web interface and `Ctrl-F`, you might need to scroll down to make GitHub load all files first.

> If you refuse to rebase

It's not that I refuse—it's the same amount of work—it's that merging means that there will be one additional commit that incorporates changes external to this PR (the 🌌 unknown 🌌), and rebasing means that the 🌌 unknown 🌌 will creep into all of the commits... While the Linux guide does the opposite of recommending it, I think it's okay to merge main into feature: if you do it last, and then fast-forward, the tree will look like a perfectly normal merge! Or I can do the merge into main on my fork, but it's really the same if I understand git right.

mikehardy commented:
I've had to workaround all sorts of stuff in ViewPager, and David's worked around quite a few things as well. In the build space gradle has issues pretty frequently but at least they iron them out mostly quickly. It's really not uncommon, unfortunately.

To the point on merging: The problem is there is no fast-forward option in github web UI so it's a maintainer pain to fast forward merge. It's possible but it can only be done via the command line and then it just "happens" to work and associate the PR and everything. So from a maintainer perspective we do rebase-merge so that the review workflow is largely web based (and thus fast and easy to distribute to new people).

I have in the past done a fast-forward merge on a couple PRs in the organization (but on the anki repository) so I know it's possible but it's a pain. I'd really prefer the rebase merge if possible 🙏

oakkitten commented:
Rebase it is, then. I tried rebasing onto the most recent tag, but that still gave conflicts, so I ended up rebasing on the current main at the time of writing.

While squashing, I took the liberty to rearrange a few things that didn't result in code changes:
  * Squashed 63ac44b18d969b20f820d2bb4898a34f02a9c05c (Add proper Javadoc for the four showSnackbar functions) into the commit introducing the code around it 
  * Same for 1a0945cd96a7e0c3d60551bb841dc2f1ec307d46 (Add tests for shouldDismiss() in SensibleSwipeDismissBehavior.kt)
  * A few minor changes to reduce diff

While reparenting, I also did a few extra changes:
  * Originally I made significant changes in `com/ichi2/anki/CoroutineHelpers.kt`, changing the two `catch` blocks into one, and showing a snackbar with indefinite length. In the upstream it has three now, and a dialog is shown for some of the errors, and a snackbar for `BackendInterruptedException`. When merging, I opted for minimal changes, as it is not clear what `BackendInterruptedException` is and why it is treated differently, and I am not sure if the errors here might contain a lot of text or new lines, which might be more suitable for a dialog. A snackbar with indefinite duration might still be a good tool for the job, still.

    Also note that my original approach was to squeeze any useful information out of an exception in a safe way, which I think is reasonable in a method like this. Now, this block says `exc.localizedMessage!!`, which may crash, or yield an empty string. I added yet another TODO about this. And by the way, `BackendInterruptedException` is defined as
    ```
    class BackendInterruptedException(error: BackendError?) : BackendException(error!!)
    ```
    which is probably an error. (Hehe, get it? An error.)
  * In `DeckPicker.kt`, I swapped `setContentView(R.layout.homescreen)` and `handleStartup()`. The latter must be not called before the former, as it may show a snackbar and that needs layout.
  * A comment somehow got all schwaggled up in Snackbars.kt, fixed that

oakkitten commented:
Rebased to resolve new conflicts

mikehardy commented:
@Arthur-Milchior I'd like to do a review on this with an eye towards merging the thing but I fear with the level of detail you got to on the review that you'll find something and I'll need to re-review and/or re-re-review etc. Could I ask you to take a look and say whether you are satisfied or not prior to me pulling the PR and doing review + testing?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -0,0 +1,138 @@
+
+enum class Dismiss { DoNotDismiss, ToTheLeft, ToTheRight }
+
+private val FLING_TO_DISMISS_SPEED_THRESHOLD =
```
same comment - AnkiDroidApp is not guaranteed

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -180,6 +183,10 @@ fun View.showSnackbar(
     snackbar.setMaxLines(2)
     snackbar.fixSwipeDismissBehavior()
 
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        snackbar.fixMarginsWhenInsetsChange()
+    }
```
nit: I feel this would be nicer if the conditional were inside the function call

oakkitten replied:
I *strongly* dislike maybe-functions. Well, in this case it could be `fixMarginsWhenInsetsChangeOnAndroidQAndAbove()` instead of `maybeFixMarginsWhenInsetsChange()`. But I'd take `if (condition) actionThatAlwaysRuns()` over this any day. Doesn't look as pretty, but easier to understand what's going on IMO.

david-allison replied:
If it's a nitpick, Implementer's choice. 

I'd have called it `fixMarginsWhenInsetsChange`, but had it be a no-op + early return if `< Q` inside the function. But: I'm fine with it the way it is, that's why it's a nitpick.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SensibleSwipeDismissBehavior.kt:
```diff
@@ -137,6 +145,33 @@ open class SensibleSwipeDismissBehavior<V : View> : SwipeDismissBehavior<V>() {
+         * When such a layout event occurs, we are undoing any horizontal changes. There's probably
+         * a more decent way of resolving this, but--again--this is an extremely rare glitch.
+         */
+        @Suppress("UNUSED_ANONYMOUS_PARAMETER")
```
nit: comment explaining that there's value in naming each `newX` parameter

oakkitten replied:
Kotlin doesn't support named ignored parameters like Python does (`_newRight`), so the only alternative is to do

```kotlin
View.OnLayoutChangeListener {
        view, newLeft, newTop, _, newBottom, oldLeft, _, oldRight, _ -> ...
}
```

I think the full version is more readable, or rather, it's easier to get what's going on. Not sure if this is worth a comment?

david-allison replied:
Nono. I agree that the annotation is the correct move from a code style perspective, but a quick comment next to the annotation explaining what it's doing would be useful

oakkitten replied:
Aha, added a comment.

david-allison reviewed (changes requested):
This looks great and I'd be glad to get this in ASAP.

Extremely well commented, and clean code.

One issue with using `AnkiDroidApp` in class/static initialisers.

The rest are nitpicks, and I'd be happy for them to either be addressed or ignored.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/FabBehavior.kt:
```diff
@@ -61,21 +63,23 @@ class FabBehavior : CoordinatorLayout.Behavior<View> {
+ * It extends to the very bottom of the screen.
+ * We allow the fab and the snackbar to overlap a bit, so that the fab doesn't appear too high.
+ */
+private val MAX_OVERLAP = UIUtils.convertDpToPixel(10f, AnkiDroidApp.getInstance().applicationContext)
```
That's surprising, never heard of this crash and I routinely use app instance like this. Any idea how to reproduce the crash on a modern device? I tried running the command from the question and apparently it needs a token now, and my device gives me no tokens.

I'm not sure if this is worth fixing here? `AnkiDroidApp.getInstance()` is used in 84 places and the callers do no nullability checks, so I guess the app is going to crash regardless.

david-allison replied:
None of these usages of `AnkiDroidApp` were in static initialisers the last time I checked. 

This was a large cause of crashes on older devices, and I would love to remove it, but it's a price I still feel is worthwhile for stability.

oakkitten replied:
I understand, but static initializers are run when the class is loaded, which in this case would be when a snackbar is actually shown. If it was ever going to, by this time the application has already long crashed. And if it hasn't, at this point it would still crash even if variables were not set statically, as `AnkiDroidApp.getInstance()` is still null. You'd have to get context from elsewhere. And even if you did that, the app would then still crash later on because `AnkiDroidApp.getInstance()` is used in 84 places... 😅

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/SwipeDismissBehaviorFix.kt:
```diff
@@ -0,0 +1,84 @@
+        override fun onShown(snackbar: Snackbar) {
+            super.onShown(snackbar)
+            val params = snackbar.view.layoutParams
+            if (params is CoordinatorLayout.LayoutParams) {
```
This gets refactored in later commits into something else. I can remove it, but I prefer to keep solutions, even if suboptimal, in git history, so I can revisit them later.

david-allison replied:
I'm fine with that

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/snackbar/Snackbars.kt:
```diff
@@ -175,14 +177,41 @@ fun View.showSnackbar(
+    get() {
+        return (view.layoutParams as? CoordinatorLayout.LayoutParams)
+            ?.behavior as? SwipeDismissBehavior
+    }
```
This also gets refactored in later commits.

(I'm not particular about `get() =` and `get() {}`, it's just that I think that here the old school version look prettier next to `set()` which needs the curly braces.)

oakkitten commented:
Updated to remove a comment and ignore in dimens.xml as requested, also threw in a `@VisibleForTesting` that I somehow forgot to put in.

mikehardy commented:
Arrrrgh - clicked over here for the umpteenth time after seeing get a shout in #12040 and seeing that there was activity, and there are conflicts again. Really difficult PR from perspective of conflicts, you have my gratitude @oakkitten for keeping up with it. Hopefully merged + put to bed soon!

oakkitten commented:
Rebased onto v2.16alpha78.

david-allison reviewed (approved):
If you've confirmed that the variables are only used when a snackbar is flung, I'm happy. Thanks!

mikehardy reviewed (approved):
These look and "feel" (behave well both non-interactively and while flinging them) under test. The commit saga shows the struggle is real, I think this is a ninja-level yak shave and I've been sad watching it take conflicts over and over. Looking forward to seeing some of the toasts converted as mentioned in a separate issue just recently. Let's go!
